File: app/src/main/java/com/huck/biblequiz/ui/screens/study/StudyViewModel.kt
package com.huck.biblequiz.ui.screens.study

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.huck.biblequiz.data.repository.BibleRepository
import com.huck.biblequiz.data.repository.ProgressRepository
import com.huck.biblequiz.model.Verse
import com.huck.biblequiz.util.BookNames
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ChapterContent(
    val bookId: Int,
    val chapter: Int,
    val bookName: String,
    val verses: List<Verse>
)

data class StudyState(
    val chapters: List<ChapterContent> = emptyList(),
    val currentIndex: Int = 0,
    val isLoading: Boolean = true,
    val error: String? = null
) {
    val currentChapter: ChapterContent? get() = chapters.getOrNull(currentIndex)
    val hasPrev: Boolean get() = currentIndex > 0
    val hasNext: Boolean get() = currentIndex < chapters.size - 1
}

@HiltViewModel
class StudyViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val bibleRepository: BibleRepository,
    private val progressRepository: ProgressRepository
) : ViewModel() {

    private val _state = MutableStateFlow(StudyState())
    val state: StateFlow<StudyState> = _state

    init {
        val selections = savedStateHandle.get<String>("selections") ?: ""
        loadChapters(selections)
    }

    private fun loadChapters(selections: String) {
        viewModelScope.launch {
            try {
                val pairs = selections.split(",").mapNotNull { s ->
                    val parts = s.split(":")
                    if (parts.size == 2) Pair(parts[0].toInt(), parts[1].toInt()) else null
                }

                val chapters = pairs.map { (bookId, chapter) ->
                    val verses = bibleRepository.getChapterVerses(bookId, chapter)
                    progressRepository.markStudied(bookId, chapter)
                    ChapterContent(
                        bookId = bookId,
                        chapter = chapter,
                        bookName = BookNames.getShortName(bookId),
                        verses = verses.map { Verse(it.bookId, it.chapter, it.verse, it.text) }
                    )
                }

                _state.value = _state.value.copy(chapters = chapters, isLoading = false)
            } catch (e: Exception) {
                _state.value = _state.value.copy(error = e.message, isLoading = false)
            }
        }
    }

    fun nextChapter() {
        if (_state.value.hasNext) {
            _state.value = _state.value.copy(currentIndex = _state.value.currentIndex + 1)
        }
    }

    fun prevChapter() {
        if (_state.value.hasPrev) {
            _state.value = _state.value.copy(currentIndex = _state.value.currentIndex - 1)
        }
    }
}

File: ios/BeSpecific/Views/Study/StudyViewModel.swift
import Foundation

struct ChapterContent {
    let bookId: Int
    let chapter: Int
    let bookName: String
    let verses: [Verse]
}

@MainActor @Observable
final class StudyViewModel {
    var chapters: [ChapterContent] = []
    var currentIndex = 0
    var isLoading = true
    var error: String?

    var currentChapter: ChapterContent? { chapters.indices.contains(currentIndex) ? chapters[currentIndex] : nil }
    var hasPrev: Bool { currentIndex > 0 }
    var hasNext: Bool { currentIndex < chapters.count - 1 }

    private let bibleRepository: BibleRepository
    private let progressRepository: ProgressRepository

    init(selections: String, bibleRepository: BibleRepository, progressRepository: ProgressRepository) {
        self.bibleRepository = bibleRepository
        self.progressRepository = progressRepository
        Task { await loadChapters(selections: selections) }
    }

    private func loadChapters(selections: String) async {
        do {
            let pairs = selections.split(separator: ",").compactMap { s -> (Int, Int)? in
                let parts = s.split(separator: ":")
                guard parts.count == 2, let b = Int(parts[0]), let c = Int(parts[1]) else { return nil }
                return (b, c)
            }

            var result: [ChapterContent] = []
            for (bookId, chapter) in pairs {
                let entities = try await bibleRepository.getChapterVerses(bookId: bookId, chapter: chapter)
                try progressRepository.markStudied(bookId: bookId, chapter: chapter)
                result.append(ChapterContent(
                    bookId: bookId,
                    chapter: chapter,
                    bookName: BookNames.getShortName(bookId),
                    verses: entities.map { Verse(bookId: $0.bookId, chapter: $0.chapter, verse: $0.verse, text: $0.text) }
                ))
            }
            chapters = result
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }

    func nextChapter() {
        if hasNext { currentIndex += 1 }
    }

    func prevChapter() {
        if hasPrev { currentIndex -= 1 }
    }
}
